cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationHelper.java

```diff
@@ -0,0 +1,101 @@
+import io.strimzi.operator.common.operator.resource.SecretOperator;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.Config;
+import org.apache.kafka.clients.admin.DescribeConfigsResult;
+import org.apache.kafka.common.config.ConfigResource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * This class contains methods for getting current configuration from the kafka brokers asynchronously.
+ */
+public class KafkaBrokerConfigurationHelper {
+
+    private final AdminClientProvider adminClientProvider;
+    protected final SecretOperator secretOperations;
+    private static final Logger log = LogManager.getLogger(KafkaBrokerConfigurationHelper.class);
+
+    public KafkaBrokerConfigurationHelper(AdminClientProvider adminClientProvider, SecretOperator secretOperations) {
+        this.adminClientProvider = adminClientProvider;
+        this.secretOperations = secretOperations;
+    }
+
+    public Future<Admin> adminClient(StatefulSet sts, int podId) {
+        Promise<Admin> acPromise = Promise.promise();
+        String cluster = sts.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
+        String namespace = sts.getMetadata().getNamespace();
+        Future<Secret> clusterCaKeySecretFuture = secretOperations.getAsync(
+                namespace, KafkaResources.clusterCaCertificateSecretName(cluster));
+        Future<Secret> coKeySecretFuture = secretOperations.getAsync(
+                namespace, ClusterOperator.secretName(cluster));
+        String hostname = KafkaCluster.podDnsName(namespace, cluster, KafkaCluster.kafkaPodName(cluster, podId)) + ":" + KafkaCluster.REPLICATION_PORT;
+
+        return CompositeFuture.join(clusterCaKeySecretFuture, coKeySecretFuture).compose(compositeFuture -> {
+            Secret clusterCaKeySecret = compositeFuture.resultAt(0);
+            if (clusterCaKeySecret == null) {
+                return Future.failedFuture(Util.missingSecretException(namespace, KafkaCluster.clusterCaKeySecretName(cluster)));
+            }
+            Secret coKeySecret = compositeFuture.resultAt(1);
+            if (coKeySecret == null) {
+                return Future.failedFuture(Util.missingSecretException(namespace, ClusterOperator.secretName(cluster)));
+            }
+
+            Admin ac;
+            try {
+                log.info("about to create an admin client");
+                ac = adminClientProvider.createAdminClient(hostname, clusterCaKeySecret, coKeySecret, "cluster-operator");
+                acPromise.complete(ac);
+            } catch (RuntimeException e) {
```

Comment:
Does this actually connect? Or when / why would this happen?

Reply:
Yeah, we should at least document this behaviour in the `AdminClientProvider`.